keps/sig-apps/20200529-delete-priority-annotations.md

```diff
@@ -0,0 +1,291 @@
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This feature allows making a suggestion to the ReplicaSet controller, which pod of a Deployment should be deleted first when a scale-down event happens. This can prevent session disruption in stateful applications in a trivial manner.
+
+## Motivation
+
+For some applications, it is necessary that the application can tell Kubernetes which pod can be deleted and which replica has to be protected. The reason for this is that some applications do have stateful sessions and it is not possible to put such an application into Kubernetes because of session termination resulting from "random" down-scale. If the application is able to tell Kubernetes which of the replicas contains no/few/less important active sessions, this would solve many problems. This feature is non-disruptive to the default behaviour. Only if the annotation is existing, it will make a difference in deletion order.
+
+### Goals
+
+To recommend which pod gets deleted next of a ReplicaSet. This should help to avoid major reworks in existing applications architecture:
+* [45509](https://github.com/kubernetes/kubernetes/issues/45509) - Scale down a deployment by removing specific pods
+
+
+### Non-Goals
+
+Guaranteed (in contrast to the recommendation stated in Goals) deletion of a selected replica.
+
+## Proposal
+
+The application can set the `controller.kubernetes.io/replicaset-downscale-priority` annotation to a pod through the Kubernetes API. When a downscale event happens, the pod with the lower priority value of the previously set annotation will be deleted first. If one pod of the Deployment has no priority annotation set, it will be treated as the lowest priority.
```

Comment:
What's a valid priority value?

Reply:
All integer values are accepted. A higher positive number describes a lower priority. The highest priority would be -MAXINT and the lowest priority would be +MAXINT

Reply:
>  A higher positive number describes a lower priority.

It'd be more intuitive if higher priority value describes a higher priority, and would be more aligned with [pod priority](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/).

Reply:
The order (-MAXINT=highest, +MAXINT=lowest) matches the order for Linux kernel niceness for Linux processes. Of course we can flip the priority order.
What do you think is better?

Reply:
I think we should be consistent with pod priority

Reply:
I suggest the opposite and second what @drbugfinder-work says. Reverting may not feel intuitive but it's easier to implement. Examples of ranking:

#### Balancing duplicates among topological domains

- Evict pods while minimizing skew between topological domains
- Each pod can be given a cost based on how old/young it is in the same domain:
  - if a pod is the first one in the domain, rank the pod with cost `1`
  - if a pod was created second to the domain, rank the pod with cost `2`
  - continue this way until all pods in all domains are ranked
  - higher rank of a pod, the sooner the pod gets removed

#### Pods not tolerating taints first

- Evict pods that do not tolerate taints before pods that tolerate taints.
- Each pod can be given a cost based on how many taints are not tolerated
  - higher rank of a pod, the sooner the pod gets removed

If we decide to be consistent with the pod priority, we will need to revert the ranks which will require specify a maximum ranking. Which can be MAXINT. However, instead of having ranking (1,2,3,...) we will get (MAXINT-1,MAXINT-2,MAXINT-3,...) which will produce hard to read and long number.

Reply:
I agree, and that is really what I meant, a higher number means higher cost/delete-priority. As you suggested in the other comment, we should depart from the name "priority" to avoid any confusion.

We should also define a default value in case some pods have this annotation set and some don't. I guess zero should work.

Reply:
In my opinion we do not need a default value. The default is that no annotation exists. In case if a user actively makes use of the delete priority annotation, it would mean that the pod _does_ have a priority over the non-annotated pod. As described, the logic would be non-annotated < low-prio annotation < high-prio annotation. If possible I would also leave any fixed priority ranges - so that there is the most flexibility (e.g. you have a pod with "prio 1000" and if the range ends with 1000, you cannot go further in case you need values behind 1000). 
However we could add some recommendation in the description, that values **should** start with 0 and go negative and positive or similar.